pool/pool.go

```diff
@@ -0,0 +1,87 @@
+package pool
+
+import "github.com/ubclaunchpad/cumulus/blockchain"
+
+// Pool is a set of valid Transactions.
+type Pool struct {
+	OrderMap        map[blockchain.Hash]int
+	OrderReverseMap map[int]blockchain.Hash
+	Transactions    map[blockchain.Hash]*blockchain.Transaction
+}
+
+// New initializes a new pool.
+func New() *Pool {
+	return &Pool{
+		OrderMap:        map[blockchain.Hash]int{},
+		OrderReverseMap: map[int]blockchain.Hash{},
+		Transactions:    map[blockchain.Hash]*blockchain.Transaction{},
+	}
+}
+
+// Len returns the number of transactions in the Pool.
+func (p *Pool) Len() int {
+	return len(p.Transactions)
+}
+
+// Get returns the transction with input transaction Hash h.
+func (p *Pool) Get(h blockchain.Hash) *blockchain.Transaction {
+	return p.Transactions[h]
+}
+
+// GetN returns the Nth transaction in the pool.
+func (p *Pool) GetN(N int) *blockchain.Transaction {
+	return p.Transactions[p.OrderReverseMap[N]]
+}
+
+// Set inserts a transaction into the pool, returning
+// true if the Transaction was inserted (was valid).
+func (p *Pool) Set(t *blockchain.Transaction, bc *blockchain.BlockChain) bool {
+	ok, _ := bc.ValidTransaction(t)
+	if ok {
+		p.OrderMap[t.Input.Hash] = p.Len()
+		p.OrderReverseMap[p.Len()] = t.Input.Hash
+		p.Transactions[t.Input.Hash] = t
+	}
+	return ok
+}
+
+// Delete removes a transaction from the Pool.
+func (p *Pool) Delete(t *blockchain.Transaction) {
+	i := p.OrderMap[t.Input.Hash]
+	if i < p.Len() {
```

Comment:
Doesn't this also need to shuffle transaction hash indices
Ex: you have `{1: a, 2: b, 3: c, 4: d, 5: e}` and remove 3 to get `{1: a, 2: b, 4: d, 5: e}`. But now d isn't the 4th element anymore it's the 3rd, same for e. Does this matter?